logger/analytics/analyticslogger.go

```diff
@@ -148,21 +175,28 @@ func (al *Logger) Write(bs []byte) (int, error) {
 	al.mu.Lock()
 	al.batchBytes += len(bs)
 	al.batch = append(al.batch, &firehose.Record{Data: bs})
-	if al.batchOldestTime == nil {
-		al.batchOldestTime = aws.Time(time.Now())
-	}
 	shouldSendBatch := len(al.batch) == al.maxBatchRecords ||
-		al.batchBytes > int(0.9*float64(al.maxBatchBytes)) ||
-		(al.maxBatchTime != 0 && time.Now().After(al.batchOldestTime.Add(al.maxBatchTime)))
+		al.batchBytes > int(0.9*float64(al.maxBatchBytes))
+	al.mu.Unlock()
+
 	if shouldSendBatch {
+		al.flush()
+	}
+	return len(bs), nil
+}
+
+// flush asynchronously flushes a batch to kinesis
+func (al *Logger) flush() error {
+	al.mu.Lock()
+	defer al.mu.Unlock()
+	if len(al.batch) > 0 {
 		batch := al.batch
-		al.batch = []*firehose.Record{}
+		al.batch = nil
 		al.batchBytes = 0
-		al.batchOldestTime = nil
 		// be careful not to send al.batch, since we will unlock before we finish sending the batch
 		al.sendBatchWG.Add(1)
 		go func() {
-			err = sendBatch(batch, al.fhAPI, al.fhStream, time.Now().Add(timeoutForSendingBatches))
+			err := sendBatch(batch, al.fhAPI, al.fhStream, time.Now().Add(timeoutForSendingBatches))
```

Comment:
Should the `.Done()` on the line below this one be `defer`ed?

Reply:
Sure, I suppose so. I guess the concern being that otherwise the main function could exit before that log gets sent?

I'll update it

Reply:
I guess the concern is that if there is a panic in `sendBatch` and the upstream does `recover()`, the waitGroup might be broken (not exactly sure how panic+recover works). I don't think there's likely to be a problem in practice, so this seems fine.

Reply:
Oh I see, yeah seems like a good justification to defer instead of just putting it at the end (better to stick to good practice even if it isn't likely to be an issue).

I'll change that before merging 👍